src/stable_marriage.rs

```diff
@@ -1,96 +1,234 @@
-struct Individual {
-    cnt_match: Option<usize>,
+use itertools::Itertools;
+use std;
+use std::cmp::Ordering;
 
-    // For efficiency: most preferred last for men, most preferred first for women
-    is_man: bool,
-    prefs: Vec<usize>,
-}
+pub trait Matcher {
+    type Item;
+    type Target;
 
-impl Individual {
-    fn get_pref_idx(&self, idx: usize) -> Option<usize> {
-        debug_assert!(!self.is_man);
-        self.prefs.iter().position(|i| *i == idx)
-    }
+    fn is_admissible(&self, x: &Self::Item, y: &Self::Target) -> bool;
 
-    fn pop_preferred_if_unmatched(&mut self) -> Option<usize> {
-        debug_assert!(self.is_man);
-        if self.cnt_match.is_some() {
-            None
-        } else {
-            self.prefs.pop()
-        }
+    fn cmp_3way(&self, from: &Self::Item, left: &Self::Target, right: &Self::Target) -> Ordering;
+
+    fn compute_preference_list<Q>(
+        &self,
+        item: &Self::Item,
+        targets: &Vec<Woman<Self>>,
+        other_matcher: &Q,
+    ) -> Vec<usize>
+    where
+        Q: Matcher<Item = Self::Target, Target = Self::Item>,
+    {
+        let mut admissibles = targets
+            .iter()
+            .enumerate()
+            .filter(|(_, x)| x.prefers_to_current(other_matcher, item))
+            .map(|(i, x)| (i, &x.data))
+            .filter(|(_, x)| self.is_admissible(item, x))
+            .collect::<Vec<_>>();
+
+        admissibles.sort_unstable_by(|(i, left), (j, right)| {
+            self.cmp_3way(item, left, right).then(i.cmp(&j)).reverse()
+        });
+
+        admissibles.into_iter().map(|(i, _)| i).collect::<Vec<_>>()
     }
-}
 
-fn make_empty_matching(v: Vec<Vec<usize>>, is_man: bool) -> Vec<Individual> {
-    v.into_iter()
-        .map(|prefs| Individual {
-            cnt_match: None,
-            is_man: is_man,
-            prefs: prefs,
-        })
-        .collect()
-}
+    // Captures a notion of a couple being made for each other.
+    // If a good number of matchings have a perfect match, quadratic behaviour is strongly reduced.
+    fn is_perfect_match(&self, _x: &Self::Item, _y: &Self::Target) -> bool {
+        false
+    }
 
-fn init_matching(
-    men: Vec<Vec<usize>>,
-    women: Vec<Vec<usize>>,
-) -> (Vec<Individual>, Vec<Individual>) {
-    let mut men = make_empty_matching(men, true);
-    let women = make_empty_matching(women, false);
-    for mut x in &mut men {
-        x.prefs.reverse();
+    fn find_perfect_match<'a>(
+        &self,
+        item: &Self::Item,
+        targets: &'a mut Vec<Woman<Self>>,
+    ) -> Option<&'a mut Woman<Self>> {
+        targets
+            .iter_mut()
+            .filter(|x| x.current_match.is_none())
+            .find(|x| self.is_perfect_match(item, &x.data))
     }
+}
 
-    (men, women)
+struct Man<P: Matcher + ?Sized> {
```

Comment:
I *knew* you'd give in and eventually have separate types for men and women :stuck_out_tongue_closed_eyes:

Reply:
Indeed ^^